### PR TITLE
Support downloading all testcases in AWS

### PR DESCRIPTION
**Pull request by myungwoo**
Add feature for #572 .

URL to download all testcases is,
"/dataset/(dataset_id)/testcases/download" .

[Review on Reviewable](https://reviewable.io/reviews/cms-dev/cms/573)

### CODE REVIEW

**Comment by stefano-maggiolo** [comment]
Reviewed 4 of 5 files at r1, 1 of 1 files at r2.
Review status: all files reviewed at latest revision, 10 unresolved discussions.

---

*[cms/server/util.py, line 579 \[r2\]](https://reviewable.io:443/reviews/cms-dev/cms/573#-KCfJmiuXziXgVEUoBOt:-KCfJmiuXziXgVEUoBOu:-1118605804) ([raw file](https://github.com/cms-dev/cms/blob/039ed633c83befe61dc8a86bd46e14c07af41072/cms/server/util.py#L579)):*
Please write something like `"""Send a file from FileCacher by its digest."""` (for single line pydoc it is ok to avoid the next empty line).

---

*[cms/server/util.py, line 597 \[r2\]](https://reviewable.io:443/reviews/cms-dev/cms/573#-KCfKAV3hUzCprTrTxLB:-KCfKAV49BSNfiLtEKCG:-273167927) ([raw file](https://github.com/cms-dev/cms/blob/039ed633c83befe61dc8a86bd46e14c07af41072/cms/server/util.py#L597)):*
Same here (one line is good, and use imperative, `Send [...]`). 

---

*[cms/server/util.py, line 602 \[r2\]](https://reviewable.io:443/reviews/cms-dev/cms/573#-KCfOsdfZ2HAKGaDjw3W:-KCfOsdgbo47X9MpSx10:-2067698544) ([raw file](https://github.com/cms-dev/cms/blob/039ed633c83befe61dc8a86bd46e14c07af41072/cms/server/util.py#L602)):*
Fits in one line

---

*[cms/server/util.py, line 605 \[r2\]](https://reviewable.io:443/reviews/cms-dev/cms/573#-KCfP738NzPToah5imiM:-KCfP738NzPToah5imiN:1353526822) ([raw file](https://github.com/cms-dev/cms/blob/039ed633c83befe61dc8a86bd46e14c07af41072/cms/server/util.py#L605)):*
Prefer this (if you want, you can change it in the other one as well):

```python
except Exception:
    logger.error("Exception while retrieving file `%s'.", filepath,
                 exc_info=True)
```

---

*[cms/server/admin/handlers/dataset.py, line 40 \[r2\]](https://reviewable.io:443/reviews/cms-dev/cms/573#-KCfPnaGDtxx9-Nrq0ku:-KCfPnaHpqcT6hv8-kDH:-1974140968) ([raw file](https://github.com/cms-dev/cms/blob/039ed633c83befe61dc8a86bd46e14c07af41072/cms/server/admin/handlers/dataset.py#L40)):*
Please keep these imports sorted.

---

*[cms/server/admin/handlers/dataset.py, line 664 \[r2\]](https://reviewable.io:443/reviews/cms-dev/cms/573#-KCfPvHQwueRqSqoJfKq:-KCfPvHQwueRqSqoJfKr:-2017558144) ([raw file](https://github.com/cms-dev/cms/blob/039ed633c83befe61dc8a86bd46e14c07af41072/cms/server/admin/handlers/dataset.py#L664)):*
Nit: "... testcases in a zip file."

---

*[cms/server/admin/handlers/dataset.py, line 683 \[r2\]](https://reviewable.io:443/reviews/cms-dev/cms/573#-KCfQI2_wtgvXVw37EG2:-KCfQI2_wtgvXVw37EG3:1032801925) ([raw file](https://github.com/cms-dev/cms/blob/039ed633c83befe61dc8a86bd46e14c07af41072/cms/server/admin/handlers/dataset.py#L683)):*
Wrap line

---

*[cms/server/admin/handlers/dataset.py, line 692 \[r2\]](https://reviewable.io:443/reviews/cms-dev/cms/573#-KCfR6D4if42t2iO3i0Y:-KCfR6D4if42t2iO3i0Z:1924771060) ([raw file](https://github.com/cms-dev/cms/blob/039ed633c83befe61dc8a86bd46e14c07af41072/cms/server/admin/handlers/dataset.py#L692)):*
The second argument to get_argument is the default value, so you can use something like `zip_filename = self.get_argument("zip_filename", "testcases.zip")`

---

*[cms/server/admin/handlers/dataset.py, line 707 \[r2\]](https://reviewable.io:443/reviews/cms-dev/cms/573#-KCfRfJgERWiGvlDnQgO:-KCfRfJgERWiGvlDnQgP:-1051750311) ([raw file](https://github.com/cms-dev/cms/blob/039ed633c83befe61dc8a86bd46e14c07af41072/cms/server/admin/handlers/dataset.py#L707)):*
We have cmscommon.archive which is supposed to handle these things, but we do not have a method that does what you want. Maybe add something there, like `create_from_filenames()`? patoolib already handles this. 

---

*[cms/server/admin/templates/download_testcases.html, line 10 \[r2\]](https://reviewable.io:443/reviews/cms-dev/cms/573#-KCfkGlhdNs4qQtWH9kw:-KCfkGlhdNs4qQtWH9kx:-1615115799) ([raw file](https://github.com/cms-dev/cms/blob/039ed633c83befe61dc8a86bd46e14c07af41072/cms/server/admin/templates/download_testcases.html#L10)):*
Since we're going to use patoolib, you don't need to mandate the format - IIRC when you give it a .tar.gz extension, for example, it automatically create the archive in the right format.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cms-dev/cms/573)*

File: cms/server/util.py [hunk 1]
```diff
@@ -585,11 +585,27 @@ def fetch(self, digest, content_type, filename):
                 self.temp_file = \
                     self.application.service.file_cacher.get_file(digest)
```
Comment (stefano-maggiolo) [inline cms/server/util.py]: Codacy Issue found: [Unused variable 'error'](https://www.codacy.com/app/s-maggiolo/cms/file/2230397155/issues/source?bid=3167167&fileBranchId=3167177#l587)

**Comment by myungwoo** [comment]
Review status: 3 of 5 files reviewed at latest revision, 11 unresolved discussions.

---

*[cms/server/admin/handlers/dataset.py, line 692 \[r2\]](https://reviewable.io:443/reviews/cms-dev/cms/573#-KCfR6D4if42t2iO3i0Y:-KChI8fsWB6NZb8ydhk7:374647716) ([raw file](https://github.com/cms-dev/cms/blob/039ed633c83befe61dc8a86bd46e14c07af41072/cms/server/admin/handlers/dataset.py#L692)):*
I also think it is much better method.
How about the same function call in AddTestcasesHandler?

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cms-dev/cms/573)*

**Comment by myungwoo** [comment]
Review status: 3 of 5 files reviewed at latest revision, 11 unresolved discussions.

---

*[cms/server/admin/handlers/dataset.py, line 707 \[r2\]](https://reviewable.io:443/reviews/cms-dev/cms/573#-KCfRfJgERWiGvlDnQgO:-KChJmOesZs9ylCqJbZC:1246342773) ([raw file](https://github.com/cms-dev/cms/blob/039ed633c83befe61dc8a86bd46e14c07af41072/cms/server/admin/handlers/dataset.py#L707)):*
I've tried to use cmscommon.archive, however I think it needs more operations.
The reason is that if I want to archive file named "cache/AdminWebServer-0/(sha1sum)" to name "input.001", it needs to copy the file to temp dir with name "input.001" instead of archiving directly.
Zipfile can handle this directly.

Without supporting tar.gz, I think it is fine because Add multiple testcasess also supports zip file only.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cms-dev/cms/573)*

**Comment by myungwoo** [comment]
Review status: 3 of 5 files reviewed at latest revision, 11 unresolved discussions.

---

*[cms/server/util.py, line 579 \[r2\]](https://reviewable.io:443/reviews/cms-dev/cms/573#-KCfJmiuXziXgVEUoBOt:-KChLBXsMuW3-8aUg7vP:-499146009) ([raw file](https://github.com/cms-dev/cms/blob/039ed633c83befe61dc8a86bd46e14c07af41072/cms/server/util.py#L579)):*
Done.

---

*[cms/server/util.py, line 597 \[r2\]](https://reviewable.io:443/reviews/cms-dev/cms/573#-KCfKAV3hUzCprTrTxLB:-KChL9JNG2_79_f3TZpm:-499146009) ([raw file](https://github.com/cms-dev/cms/blob/039ed633c83befe61dc8a86bd46e14c07af41072/cms/server/util.py#L597)):*
Done.

---

*[cms/server/util.py, line 602 \[r2\]](https://reviewable.io:443/reviews/cms-dev/cms/573#-KCfOsdfZ2HAKGaDjw3W:-KChLCPdWT4CQKxsXqw-:-499146009) ([raw file](https://github.com/cms-dev/cms/blob/039ed633c83befe61dc8a86bd46e14c07af41072/cms/server/util.py#L602)):*
Done.

---

*[cms/server/util.py, line 605 \[r2\]](https://reviewable.io:443/reviews/cms-dev/cms/573#-KCfP738NzPToah5imiM:-KChLCwPunLwBg2MCm-P:-499146009) ([raw file](https://github.com/cms-dev/cms/blob/039ed633c83befe61dc8a86bd46e14c07af41072/cms/server/util.py#L605)):*
Done.

---

*[cms/server/util.py, line 586 \[r3\]](https://reviewable.io:443/reviews/cms-dev/cms/573#-KBuwP0z17e1gmTbAgb0-r3-586:-KChLAceycPrC7aRvtos:-499146009) ([raw file](https://github.com/cms-dev/cms/blob/1873f98b00f2f4166612faab69977676f5ba8e65/cms/server/util.py#L586)):*
Done.

---

*[cms/server/admin/handlers/dataset.py, line 40 \[r2\]](https://reviewable.io:443/reviews/cms-dev/cms/573#-KCfPnaGDtxx9-Nrq0ku:-KChLDW7CXip6Y6YeLsd:-499146009) ([raw file](https://github.com/cms-dev/cms/blob/039ed633c83befe61dc8a86bd46e14c07af41072/cms/server/admin/handlers/dataset.py#L40)):*
Done.

---

*[cms/server/admin/handlers/dataset.py, line 664 \[r2\]](https://reviewable.io:443/reviews/cms-dev/cms/573#-KCfPvHQwueRqSqoJfKq:-KChLE5uy4qkXRL1P5up:-499146009) ([raw file](https://github.com/cms-dev/cms/blob/039ed633c83befe61dc8a86bd46e14c07af41072/cms/server/admin/handlers/dataset.py#L664)):*
Done.

---

*[cms/server/admin/handlers/dataset.py, line 683 \[r2\]](https://reviewable.io:443/reviews/cms-dev/cms/573#-KCfQI2_wtgvXVw37EG2:-KChLEefq5NDdE9G_TF7:-499146009) ([raw file](https://github.com/cms-dev/cms/blob/039ed633c83befe61dc8a86bd46e14c07af41072/cms/server/admin/handlers/dataset.py#L683)):*
Done.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cms-dev/cms/573)*

**Comment by stefano-maggiolo** [comment]
Reviewed 1 of 2 files at r3, 1 of 1 files at r4.
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[cms/server/admin/handlers/dataset.py, line 692 \[r2\]](https://reviewable.io:443/reviews/cms-dev/cms/573#-KCfR6D4if42t2iO3i0Y:-KCjN1Z6W_i4IaMyUTS6:605090013) ([raw file](https://github.com/cms-dev/cms/blob/039ed633c83befe61dc8a86bd46e14c07af41072/cms/server/admin/handlers/dataset.py#L692)):*
Yep, good catch! Feel free to change there as well.

---

*[cms/server/admin/handlers/dataset.py, line 707 \[r2\]](https://reviewable.io:443/reviews/cms-dev/cms/573#-KCfRfJgERWiGvlDnQgO:-KCjNAJguXN3cddKpZrA:1982899049) ([raw file](https://github.com/cms-dev/cms/blob/039ed633c83befe61dc8a86bd46e14c07af41072/cms/server/admin/handlers/dataset.py#L707)):*
Right =)
Ok, no problem.

---

*[cms/server/admin/handlers/dataset.py, line 709 \[r4\]](https://reviewable.io:443/reviews/cms-dev/cms/573#-KCjNPBg0YrVlie71zxo:-KCjNPBg0YrVlie71zxp:1444506891) ([raw file](https://github.com/cms-dev/cms/blob/d39f3f8e8d4e553c6ce185f8c26744c09b5163c8/cms/server/admin/handlers/dataset.py#L709)):*
Only 4 spaces from the start of the line:
```python
input_path = ...\
    get_file...
```

also two lines later

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cms-dev/cms/573)*

**Comment by myungwoo** [comment]
Review status: 4 of 5 files reviewed at latest revision, 2 unresolved discussions.

---

*[cms/server/admin/handlers/dataset.py, line 692 \[r2\]](https://reviewable.io:443/reviews/cms-dev/cms/573#-KCfR6D4if42t2iO3i0Y:-KClD5v5DnwWEirF2Ugl:-499146009) ([raw file](https://github.com/cms-dev/cms/blob/039ed633c83befe61dc8a86bd46e14c07af41072/cms/server/admin/handlers/dataset.py#L692)):*
Done.

---

*[cms/server/admin/handlers/dataset.py, line 709 \[r4\]](https://reviewable.io:443/reviews/cms-dev/cms/573#-KCjNPBg0YrVlie71zxo:-KClD9i1LXtLbiIc9cVk:-499146009) ([raw file](https://github.com/cms-dev/cms/blob/d39f3f8e8d4e553c6ce185f8c26744c09b5163c8/cms/server/admin/handlers/dataset.py#L709)):*
Done.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cms-dev/cms/573)*

**Comment by myungwoo** [comment]
Reminder! @stefano-maggiolo 
The issues you told me are all resolved. :)

**Comment by stefano-maggiolo** [comment]
Sorry, life took over :) Looks good, will merge later today

---

Reviewed 1 of 1 files at r5.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cms-dev/cms/573#-:-KErV2djlKyz5Kx0XQod:940089930)*

**Comment by stefano-maggiolo** [comment]
Merged, just wrapped a couple of long lines and changed a variable name which should have been changed after a copy paste. Sorry it took such a long time!

**Pull request closed without merging**